ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -246,6 +250,69 @@ public boolean updateAccounts(Collection<Account> accounts) {
     return updateAccountsWithAccountMetadataStore(accounts, accountMetadataStore);
   }
 
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public Pair<String, String> addContainer(String accountName, Container newContainer) {
+    checkOpen();
+    if (accountName == null || accountName.isEmpty() || newContainer == null || newContainer.getName() == null
+        || newContainer.getName().isEmpty()) {
+      logger.error("Account name: {} or container: {} is invalid", accountName, newContainer);
+      throw new IllegalArgumentException("Account or container name is null or empty");
+    }
+    Account account = getAccountByName(accountName);
+    if (account == null) {
+      logger.error("Account {} is not found", accountName);
+      throw new IllegalArgumentException("Account " + accountName + " is not found");
+    }
+    String accountId = String.valueOf(account.getId());
+    for (Container container : account.getAllContainers()) {
+      // make sure there is no conflict container (conflict means a container with same name but different attributes already exists).
+      if (container.getName().equals(newContainer.getName())) {
+        if (containerEditableFieldsAreSame(container, newContainer)) {
+          // If an exactly same container already exists, we directly return its id. Adding same container multiple times
+          // should be no-op.
+          return new Pair<>(accountId, String.valueOf(container.getId()));
+        } else {
+          logger.error("A container: {} already exists with different attributes", container.getName());
+          throw new IllegalArgumentException("There is a conflicting container in account " + accountName);
+        }
+      }
+    }
+    // if code reaches here, it means no duplicate container in this account
+    // generate container id for new container
+    short nextContainerId = account.getAllContainers()
+        .stream()
+        .map(Container::getId)
+        .max(Short::compareTo)
+        .map(maxId -> (short) (maxId + 1))
+        .orElse(CONTAINER_ID_START_RANGE);
+    // construct a container based on input container and next containerId
+    Container containerToAdd = new ContainerBuilder(nextContainerId, newContainer.getName(), newContainer.getStatus(),
```

Comment:
`ContainerBuilder` has a copy constructor, could we use that here?

Reply:
Yep, we can. Thanks for the suggestion.